### DISCUSSION TITLE
@OneToOne 연관관계일 때 지연 로딩으로 설정하면 어떤 일이 발생할까요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#203] [Spring Data]
뭘까요

**Comment by jimi567** [comment]
@OneToOne 연관관계일 때 지연 로딩으로 설정하면 다음과 같은 상황이 발생할 수 있음을 고려해야 합니다. 

1. 세션 관리 
지연 로딩은 엔티티에 접근할 때, 세션이 열려 있어야 합니다. 만약 세션이 닫힌 상태에서 프록시 객체에 접근하려고 하면 LazyInitializationException 라는 예외가 발생할 수 있습니다.
2. [N+1문제](https://github.com/2024-woowacourse-study/level-interview/discussions/150)가 발생할 수 있습니다.

3. **Lazy로 설정해도 Eager처럼 동작할 수 있습니다.**
OneToOne 양방향 연관 관계에서 연관 관계 주인이 아닌 쪽 엔티티를 조회할 때, Lazy로 동작할 수 없습니다.

지연 로딩은 로딩되는 시점에 엔티티를 프록시 객체로 가져오게 됩니다. 하지만 프록시 객체의 경우 null로 감쌀 수 가 없습니다.   
연관 관계 주인(FK소유)가 아닌 의 엔티티의 경우 자신이 어떤 엔티티(칼럼)와 연관 관계를 가지고 있는지를 확인할 수 없습니다.  FK를 가지지 않는 쪽의 엔티티는 연관된 엔티티가 null인지 아닌지를 확인할 방법이 없으므로, Lazy여도 Eager로 동작하게 됩니다.  
  

예를 들어, 아래와 같은 엔티티가 있다고 가정하면,

```java

@Entity
public class User {
    @Id
    @GeneratedValue
    @Column(name = "USER_ID")
    private Long id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "STATUS_ID")
    private Status status;

}
@Entity
public class Status {

    @Id
    @GeneratedValue
    @Column("CART_ID")
    private Long id;

    @OneToOne(mappedBy = "status", fetch = FetchType.LAZY)
    private User user;
}
```
USER TABLE 
|ID|STATUS_ID(FK)|
|--|--|
|1|3|
|2|2|
|3|null|

STATUS TABLE
|ID|
|-|
|1|
|2|
|3|

이 상태에서 연관 관계 주인인 USER의 경우 각 칼럼 자신이 어떤 STATUS와 연관 관계를 가지고 있는지 알 수 있기 때문에 USER의 경우 Lazy로 정상 동작하지만  
STATUS를 조회할 때, 자신이 어떤 USER와 연관 관계를 가지고 있는지 확인할 수 있는 방법이 없고(null인지 아닌지 확인 못함) , 프록시 객체는 null로 감쌀 수 없기 때문에 Eager로 동작하게 됩니다.

**Comment by GIVEN53** [comment]
양방향 `@OneToOne`일 때 연관관계의 주인이 아닌 엔티티를 조회할 경우 Lazy Loading이 적용되지 않습니다.

db 테이블 관점에서 연관관계의 주인이 아닌 엔티티는 fk가 존재하지 않아서 엔티티의 존재 여부를 알지 못합니다.
따라서 JPA는 null 혹은 프록시 객체 중 어떤 객체를 넣어야 할지 알 수 없어서 연결된 엔티티가 있는지 없는지 확인하는 과정이 필요합니다 결과적으로 Eager처럼 동작해서 select 쿼리가 2번 발생합니다.

_User_
```java
@Entity(name = "users")
public class User {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne(mappedBy = "user", fetch = FetchType.LAZY)
    private Account account;

    protected User() {
    }

    public Account getAccount() {
        return account;
    }
}
```

_Account_
```java
@Entity
public class Account {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne(fetch = FetchType.LAZY)
    private User user;

    protected Account() {
    }

    public Account(User user) {
        this.user = user;
    }
}
```

_Test_
```java
@Test
void lazyTest() {
    userRepository.save(new User());

    User user = userRepository.findById(1L).orElseThrow();

    assertThat(user.getAccount()).isNull();
}
```

결과
<img width="244" alt="스크린샷 2024-06-24 오전 12 26 00" src="https://github.com/2024-woowacourse-study/level-interview/assets/101033262/0f213215-6f24-4042-9122-4b733f685cfa">


Reference : https://thorben-janssen.com/hibernate-tip-lazy-loading-one-to-one/
